### ISSUE TITLE
upgrade from 0.10.0 to 0.14.0 (or main) behavior change when time series returns no results

### ISSUE THREAD

**Issue by joemiller** [go-graphite/carbonapi#506]
Ran into this issue when upgrading from 0.10.0 to 0.14.0. Tried the latest nightly build carbonapi_0.14.0-39+sha.g983afd_amd64.deb as well.

Backend is go-carbon 0.12.0 but also saw the issue when upgrading to 0.15.0, so I think this is a behavioral change in carbonapi.

Previously, when querying a metric that exists but has no metrics in the current time period the result would be http/200 with an image rendered with the red text `No Data`. Now, however, the result is a 500 and the text `no metrics in the Response`.

**Comment by deniszh** [comment]
Hi @joemiller ,

Could you please include carbonapi config file that you are using ?

**Comment by joemiller** [comment]
```yaml
# carbonapi config file

listen: "0.0.0.0:8080"
concurency: 50
cache:
   # Type of caching. Valid: "mem", "memcache", "null"
   type: "mem"
   size_mb: 1024
   defaultTimeoutSec: 60
cpus: 8
tz: ""
sendGlobsAsIs: false
maxBatchSize: 2000
graphite:
    # Host:port where to send internal metrics
    # Empty = disabled
    host: "graphite-ingest:2003"
    interval: "60s"
    prefix: "carbon.carbonapi"
    pattern: "{prefix}.{fqdn}"
idleConnections: 100
pidFile: ""
notFoundStatusCode: 200
# See https://github.com/go-graphite/carbonzipper/blob/master/example.conf#L70-L108 for format explanation
upstreams:
    buckets: 10

    timeouts:
        global: "300s"
        afterStarted: "120s"
        connect: "500ms"

    concurrencyLimit: 0
    keepAliveInterval: "30s"
    maxIdleConnsPerHost: 100
    backends:
        - "http://go-carbon-0.go-carbon:8000"
        - "http://go-carbon-1.go-carbon:8000"
    graphite09compat: false
expireDelaySec: 10
logger:
    - logger: ""
      file: "stderr"
      encoding: "json"
      encodingTime: "iso8601"
      encodingDuration: "seconds"
```

I thought perhaps the `notFoundStatusCode: 200` might help but it did not, and looking at the code that seems to apply only in the case of no matching metric name found, and not the case i'm seeing which is a metric exists but there were no data points within the requested interval. This is common in our setup with some of the statsd metrics due to our statsd configuration which does not keep state and send 0's for infrequently updated metrics.

I'm currently bisecting through the `deb` packages to find when this changed. I think it might be in 0.13.0. So far I have tested the `0.11.0-2-g4a6f7c` and `0.12.0-3-ga99c07` packages and the expected behavior of a 200 response with a `no data` image remains in both.

**Comment by joemiller** [comment]
Unrelated to this issue and I'm not sure if this is ultimately a regression or expected change (probably depends on what the current versions of graphite-web are doing), but I noticed when a query has a filter and does not match any metrics the JSON result is changed.

Given an example query like: `currentAbove(servers.*.journal-2-logstash.seconds_behind.value,300)` + `format=json` and none of the metrics are above `300`:

Previous behavior (carbonapi 0.10.0): `[]`
carbonapi 0.11/0.12+ : `[{"target":"","datapoints":[]}]`

In the few places we have scripts fetching json data we have worked around this by being defensive and converting `[]` to the newer `[{"target":"","datapoints":[]}]` format.

**Comment by Civil** [comment]
Issue with png images for "no metrics found" case was fixed in this commit: https://github.com/go-graphite/carbonapi/commit/c28d5560103ad1fd82d238c9e3fc10cb2678a8d2

As about your second question about `currentAbove` and it's response - I would double check what graphite-web would return. However if you are parsing json with any json parser that shouldn't be a problem.

**Comment by joemiller** [comment]
@Civil I tried `0.14.0-40+sha.g07ee90_amd64.deb` but behavior is still:

1. default (image): 500/`Internal Server Error: error or no response: no metrics in the Response`
2. `format=json`: 500/`Internal Server Error: error or no response: no metrics in the Response`

**Comment by joemiller** [comment]
I'm new to the codebase so apologies up front if i'm off track...

I think the change in behavior occurred in 0.13.0 time frame when the `zipper.FetchProtoV3()` introduced this conditional that returns `ErrNoMetricsFetched`: https://github.com/go-graphite/carbonapi/blob/8b711944a24014e1e30ab8210cfe12ce6c935b74/zipper/zipper.go#L327-L339

The prior tag, 0.12.6, the same function looks like it does not return `ErrNoMetricsFetched` when no metrics are returned from the backend and (i think) simply returns an empty `stats`. https://github.com/go-graphite/carbonapi/blob/0.12.6/zipper/zipper.go#L350-L398

**Comment by Civil** [comment]
Interesting enough I cannot reproduce so far neither of your issues... Real no data returns "No data" for me and if the metric is just outside of the range it renders empty graph (as go-carbon replies with set of nulls).

However I've noticed that you have 2 backends configured and you are also using old config format (pre-0.12). So does both backends answer the same or one of them answers with 404 and another with something else?

Also debug log for the query would be really helpful at that point (I don't really care about metric names or backend names or whatever, so feel free to redact that information).

**Comment by Quarry77** [comment]
I also believe I have this issue. I'm updating from our old graphite stack to go-carbon/carbonapi. My Grafana dashboards are full of internal server error flags because all the render requests that have no data are returning "Internal Server Error: error or no response: no metrics in the Response" instead of a no data graph like they did before.

**Comment by joemiller** [comment]
@Civil Yeah, 2 go-carbon backends setup as a mirror-like configuration. That said both backends don't necessarily have the same data due to one dropping some metrics due to overflow or just missing the metrics due to a restart.

How would I directly query the go-carbon backends to see the returned data from each? Would that be by directly querying the carbonlink port on each go-carbon instance or should I use debug logging in carbonapi?

**Comment by Civil** [comment]
go-carbon talks with carbonapi through carbonserver. It can return raw data in json.

You can try querying URL that looks like that (note that go-carbon doesn't support any functions by itself, but can expand globs)
```
http://go-carbon-0.go-carbon:8000/render/?target=servers.*.journal-2-logstash.seconds_behind.value&format=json&from=$(date +%s --date "1 hour ago")&until=$(date +%s)
```

It doesn't support anything except unix time stamp in from or until and you always must specify them as it doesn't have any defaults.

Query example was from your answers above.

Also in debug log carbonapi will have the exact URL it's querying.

**Comment by joemiller** [comment]
@Civil When I use a fixed value for from/to both go-carbon backends appear to return the same data:

```console
export from=$(date +%s --date '1 hour ago')
export to=$(date +%s)

curl "http://go-carbon-0.go-carbon:8000/render/?target=servers.*.journal-2-logstash.seconds_behind.value&format=json&from=$from&until=$to" -s >/tmp/go-carbon-0
curl "http://go-carbon-1.go-carbon:8000/render/?target=servers.*.journal-2-logstash.seconds_behind.value&format=json&from=$from&until=$to" -s >/tmp/go-carbon-1

$ md5sum go-carbon-0 go-carbon-1
8aab3af46870f0880d6b06b4fb9a6063  go-carbon-0
8aab3af46870f0880d6b06b4fb9a6063  go-carbon-1
```

Working on getting carbonapi query debug logs next.

**Comment by Civil** [comment]
For the query that produce 5xx (you've mentioned it's for the metirc that exists but doesn't have any results) most important thing is what actually backed replies (how it looks on go-carbon's side).

**Comment by joemiller** [comment]
@Civil For that metric go-carbon returns 200 with 0 byte body:

```console
$ curl "http://go-carbon-0.go-carbon:8000/render/?target=stats.timers.foo.cacheserver.failure.count&format=json&from=$from&until=$to" -v

> GET /render/?target=stats.timers.foo.cacheserver.failure.count&format=json&from=1596390440&until=1596394042 HTTP/1.1
> Host: go-carbon-0.go-carbon:8000
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: application/text
< Vary: Accept-Encoding
< Date: Sun, 02 Aug 2020 18:48:54 GMT
< Content-Length: 0
<
* Curl_http_done: called premature == 0
* Connection #0 to host go-carbon-0.go-carbon left intact
```

**Comment by Civil** [comment]
Please test with the current master (commit that closed this case or newer).

I think I found the problem here:

Your request was not sending globs as-is, that causes request to be splitted into separate /metrics/find request and then /render request (even in case when there are no globs in request). Find request here would fail with "no metrics found", but the error was not sent through, so the http code (404) was lost and current logic when to return `500` or `no data` image based on http code that's set by zipper (default is 500).

**Comment by joemiller** [comment]
@Civil Yup! That looks much better now. I used the 0.14.0-44+sha.g3e522e_amd64.deb package from the autobuild repo to verify. I appreciate your efforts here, also thanks for some new debugging insights re: carbonapi / go-carbon I learned.

I can't fully switch to 0.14.x yet, however, as we have some other queries that look like they're timing out. It likely has something to do with the glob changes between 0.12-0.14, I think. I have some new ways on debugging though so I'll work on figuring out what's going on there.

thanks.
